src/launchpad/utils/treemap/dex_element_builder.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from typing import Dict, List
+
+from ...models.common import FileInfo
+from ...models.treemap import TreemapElement, TreemapType
+from ...parsers.android.dex.types import ClassDefinition
+from ...utils.logging import get_logger
+from .treemap_element_builder import TreemapElementBuilder
+
+logger = get_logger(__name__)
+
+
+class DexElementBuilder(TreemapElementBuilder):
+    def __init__(
+        self,
+        download_compression_ratio: float,
+        filesystem_block_size: int | None = None,
+        class_definitions: list[ClassDefinition] | None = None,
+    ) -> None:
+        super().__init__(
+            download_compression_ratio=download_compression_ratio,
+            filesystem_block_size=filesystem_block_size,
+        )
+        self.class_definitions = class_definitions or []
+
+    def build_element(self, file_info: FileInfo, display_name: str) -> TreemapElement | None:
+        # Skips using the file_info.path and leverages the class_definitions
+        # to build the treemap. This is because there could be multiple
+        # DEX files in APK and we want to group them by package vs file.
+
+        install_size = file_info.size
+        download_size = int(install_size * self.download_compression_ratio)
+
+        package_classes = self._group_classes_by_package()
+
+        package_elements: list[TreemapElement] = []
+        for package_name, classes in package_classes.items():
+            package_element = self._create_package_element(package_name, classes)
+            package_elements.append(package_element)
+
+        details = {
+            "fileExtension": file_info.file_type,
+            "class_count": len(self.class_definitions),
+        }
+
+        return TreemapElement(
+            name=display_name,
+            install_size=install_size,
+            download_size=download_size,
+            element_type=TreemapType.DEX_FILES,
+            path=file_info.path,
+            is_directory=True,
+            children=package_elements,
+            details=details,
+        )
+
+    def _group_classes_by_package(self) -> Dict[str, List[ClassDefinition]]:
+        package_classes: Dict[str, List[ClassDefinition]] = defaultdict(list)
+
+        for class_def in self.class_definitions:
+            package_name = self._extract_package_name(class_def)
+            package_classes[package_name].append(class_def)
+
+        return dict(package_classes)
+
+    def _extract_package_name(self, class_def: ClassDefinition) -> str:
+        fqn = class_def.fqn()
+        parts = fqn.split(".")
+        if len(parts) > 1:
+            # Join all parts except the last one (class name)
+            return ".".join(parts[:-1])
+        else:
+            raise ValueError(f"Invalid class definition: {fqn}")
+
+    def _create_package_element(self, package_name: str, classes: List[ClassDefinition]) -> TreemapElement:
```

Comment:
This is a relatively rudimentary implementation that just adds packages when there's a class within them - i.e. `com.emergetools.hackernews.MainActivity` would produce a single `com.emergetools.hackernews` package. We'll want to add packages as well for all the levels above that, like `com` then `com/emergetools`, etc. That will come in the near future as I polish up the treemap